File: 프로그래머스/3/43238. 입국심사/입국심사.cpp
#include <string>
#include <vector>
#include <algorithm>
using namespace std;
typedef long long ll;
long long solution(int n, vector<int> times) {
    ll answer = 0;
    sort(times.begin(), times.end());
    ll mx = n * (ll)times.back();
    ll mi = 0;
    while(mi <= mx){
        ll mid = (mi + mx)/ 2;
        ll temp = 0;
        for(int i = 0; i < times.size(); i++){
            temp += mid/(ll)times[i];
        }
        if(temp >= n){
            answer = mid;
            mx = mid - 1;
        }else{
            mi = mid + 1;
        }
    }
    return answer;
}

File: 프로그래머스/3/43238. 입국심사/입국심사.java
import java.util.*;

class Solution {
    public long solution(int n, int[] times) {
        long answer = 0;
        Arrays.sort(times);
        long l = 1, r = (long)times[times.length - 1] * n;
        while(l <= r){
            long mid = (l + r) / 2;
            long cnt = 0;
            for(int i = 0; i < times.length; i++){
                cnt += mid / times[i];
            }
            if(cnt >= n){
                answer = mid;
                r = mid - 1;
            }else{
                l = mid + 1;
            }
        }
        return answer;
    }
}